src/librustc_typeck/check/mod.rs

```diff
@@ -856,9 +856,8 @@ fn typeck_tables_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
             fcx
         };
 
-        fcx.select_all_obligations_and_apply_defaults();
-        fcx.closure_analyze(body);
         fcx.select_obligations_where_possible();
+        fcx.closure_analyze(body);
```

Comment:
I am pondering whether this is the right change. 

I mean, the original intention of applying fallback where we did was that -- after this point -- we were not supposed to be introducing "new constraints" that might have influenced the type of the variables. Therefore, it makes sense to do fallback (which may in turn introduce new constraints, leading to some iteration).

Do you think we are introducing new constraints ?

In general, though, I am on board for trying to restructure typeck to be as "lazy" as possible, so I am trying to decide whether this change is in fact doing that. It's not the way I had originally thought to go about things, which would be more a matter of identifying places where we invoke `structurally_resolve_type` or similar helpers and changing them into deferrable obligations (which would then be ultimately resolved during `select_obligations_where_possible`).

(Though a pre-req for doing that work is probably reworking the trait system to improve efficiency,  another effort I'd like to have started yesterday.)

Reply:
More specifically, things like closure upvar inference (and casts) were meant to operate over the "fully inferred types", basically. That said, I think there are corner cases that make this something of a fiction. Have to bring that back into cache.

Reply:
> identifying places where we invoke `structurally_resolve_type` or similar helpers and changing them into deferrable obligations

I understand this goal, in my simplistic bovine mind the ideal type checker goes like this:
1. Introduce all lazy constraints.
2. Solve as much as possible.
3. Do fallback.
4. Solve as much as possible.

This PR tries to make it blatantly obvious in the code that no other constraints are added between steps `3` and `4`. 

> Therefore, it makes sense to do fallback (which may in turn introduce new constraints, leading to some iteration).

I guess we haven't found any cases where the "early" fallback was actually helping inference more than the "late" fallback does.

> Do you think we are introducing new constraints such that fallback is occurring too early?

I gave the following example up-thread:
```rust
let x = 512;
x as u8;
```
Casting would hint `x` to `u8`, but fallback forces it to `i32`, this is the only case I found where early fallback was needed for backwards compatibility.

I don't know if it's true that we don't introduce any new constraints in `closure_analyze`, but it sounds like something difficult to reason about. With this PR we wouldn't have to reason about it anymore. That fact that this PR passes CI is empirical evidence that `closure_analyze` isn't really relying on fallback, whether that's really true is beyond my knowledge.

Reply:
> This PR tries to make it blatantly obvious in the code that no other constraints are added between steps 3 and 4.

OK, that makes sense to me.

> I guess we haven't found any cases where the "early" fallback was actually helping inference more than the "late" fallback does.

Well, consider the upvar inference in particular. This code is tasked with figuring out whether a closure ought to be `FnOnce`, `FnMut`, or `Fn`. It does this by looking at what sorts of things the closure does when it executes. For example, given this closure:

```rust
|| foo(x)
```

if the type of `x` is `Vec<String>`, that's a move of `x`, and hence the closure must be `FnOnce`. But if the type is `u32`, then this is a copy., and the closure can be simply `Fn`.

So what are we to do if the type is not yet inferred and we are later going to come and apply defaults? We can't know yet whether that is a move or a copy.

I think the only thing we could do would be to defer the decision about what traits the closure implements until after defaults have been applied.

Reply:
Based on your explanation, it seems plausible that upvar inference would like to add constraints, consider:
```rust
fn fn_closure<F: Fn()>(f: F) {}

fn main() {
    let x = None; // Something uninferred.
    fn_closure(|| std::mem::drop(x));
}
```
It seems reasonable that upvar inference would add a `Copy` bound to `x` or perhaps a [Move](https://github.com/Zoxc/rfcs/blob/static/text/0000-immovable-types) bound (though I don't see how those bounds could actually help inference).

It's also possible that defaulting would prevent upvar inference from working, in `fn foo<T: FnOnce() = Fn()>(_: &T) {}` after defaulting `T` is stuck as `Fn()` when upvar inference could have inferred it to `FnOnce()`.

So perhaps upvar inference isn't a special case and should be a lazy constraint, while it isn't it would use the usual mechanisms for eager type resolution such as `structurally_resolve_type`. Whether we want defaulting in eager type resolution is something to be decided.

Reply:
Argh, sorry it took me a few days to respond.

>  it seems plausible that upvar inference would like to add constraints, consider:

That is true, you could imagine it imposing the requirement, but it's now how it's defined today. There are actually two things involved here. First, we have a rule (which perhaps was not the wisest rule, but it's currently there) that based on the "expected type and where-clauses", we will select the closure kind -- so in this case, upvar inference actually doesn't get involved, we just pick the closure kind as `Fn` from the get go. (If you're interested, I'd like to try and experiment where we back off from that rule and measure the impact.)

In that case, the error is detected during borrowck, which comes after typeck and therefore can assume all types are fully known. Note that in general the requirement for things to be `Copy` (or else be moved) is flow dependent and is ultimately enforced on MIR, and we require all types to be built in order to build MIR. 

But if you wrote the example as:

```rust
let x = || drop(x);
fn_closure(x);
```

then the expected type is not involved, and indeed the closure would be selected as `FnOnce`, because it needs to own its content. This would result in an error as well, but later on, when invoking `fn_closure` (since `Fn` would not be satisfied).

----

> It's also possible that defaulting would prevent upvar inference from working

Actually, upvar inference is supposed to be **independent** of the demands that are placed on the closure (modulo that bit about the "expected type"). That is, we select the most permissive trait that the closure could possibly implement, and that is weighed against what it required to implement by others. 

In any case, in your example, the type variable `T` would be inferred to the closure's unique type, so the default (of `dyn Fn()`) would never be used.

----

> So perhaps upvar inference isn't a special case and should be a lazy constraint

I'm not convinced yet. =) Keep in mind that it's always something we could change later.